# مسئله پیاده‌سازی دوم

## گام 1 - پیش پردازش داده‌ها

### کتابخانه‌های مورد نیاز

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
    from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix

### خواندن دیتاست

In [2]:
data = pd.read_csv("./titanic.csv")

In [3]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
data.shape

(891, 12)

Rows/Record/Sample: 891

Column/Feature/Dimension: 12

به دلیل بی‌معنا بودن برای مدل "PassengerId", "Name", "Ticket" حذف ستون‌های 

In [5]:
# دو روش متفاوت برای حذف ستون اما خروجی یکسان

data.drop(["PassengerId", "Name", "Ticket"], axis='columns', inplace=True)
# data = data.drop(["PassengerId", "Name", "Ticket"], axis='columns')

In [6]:
data.shape # "PassengerId", "Name", "Ticket" Columns have beed deleted

(891, 9)

Rows/Record/Sample: 891

Column/Feature/Dimension: 9

In [7]:
data.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    male  22.0      1      0   7.2500   NaN        S
1         1       1  female  38.0      1      0  71.2833   C85        C
2         1       3  female  26.0      0      0   7.9250   NaN        S
3         1       1  female  35.0      1      0  53.1000  C123        S
4         0       3    male  35.0      0      0   8.0500   NaN        S

تبدیل‌ داده‌های رشته‌ای جنسیت به باینری 

In [8]:
data['Sex'].replace(['female', 'male'], [0, 1], inplace=True)

In [9]:
data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    1  22.0      1      0   7.2500   NaN        S
1         1       1    0  38.0      1      0  71.2833   C85        C
2         1       3    0  26.0      0      0   7.9250   NaN        S
3         1       1    0  35.0      1      0  53.1000  C123        S
4         0       3    1  35.0      0      0   8.0500   NaN        S

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int64  
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 62.8+ KB


In [11]:
data.describe()

Survived      Pclass         Sex         Age       SibSp       Parch  \
count  891.000000  891.000000  891.000000  714.000000  891.000000  891.000000   
mean     0.383838    2.308642    0.647587   29.699118    0.523008    0.381594   
std      0.486592    0.836071    0.477990   14.526497    1.102743    0.806057   
min      0.000000    1.000000    0.000000    0.420000    0.000000    0.000000   
25%      0.000000    2.000000    0.000000   20.125000    0.000000    0.000000   
50%      0.000000    3.000000    1.000000   28.000000    0.000000    0.000000   
75%      1.000000    3.000000    1.000000   38.000000    1.000000    0.000000   
max      1.000000    3.000000    1.000000   80.000000    8.000000    6.000000   

             Fare  
count  891.000000  
mean    32.204208  
std     49.693429  
min      0.000000  
25%      7.910400  
50%     14.454200  
75%     31.000000  
max    512.329200

حساب کردن تعداد مقادیر خالی

In [12]:
data.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Cabin       687
Embarked      2
dtype: int64

به دلیل زیاد بودن تعداد مقادیر خالی این ستون، بهتر است حذف شود

In [13]:
data.drop("Cabin", axis='columns', inplace=True)

In [14]:
data.shape

(891, 8)

Rows/Record/Sample: 891

Column/Feature/Dimension: 8

In [15]:
data.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [16]:
data['Age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

پر کردن مقادیر خالی ستون با استفاده از میانگین ستون

In [17]:
data['Age'].fillna((data['Age'].mean()), inplace=True)

In [18]:
data.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    2
dtype: int64

In [19]:
data['Embarked'].describe()

count     889
unique      3
top         S
freq      644
Name: Embarked, dtype: object

این ستون دارای ۳ مقدار یکتا می‌باشد

In [20]:
data['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

پر کردن مقادیر خالی این ستون با بیشترین مقدار تکرار از این ستون

In [21]:
data['Embarked'].fillna("S", inplace=True)

داده‌های خالی پر شدند

In [22]:
data.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [23]:
data.shape

(891, 8)

In [24]:
le = LabelEncoder()

تبدیل داده‌های دسته‌بندی به داده‌های عددی با استفاده از کتابخانه

In [25]:
data['Embarked'] = le.fit_transform(data['Embarked'])

In [26]:
le.classes_

array(['C', 'Q', 'S'], dtype=object)

In [27]:
data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    1  22.0      1      0   7.2500         2
1         1       1    0  38.0      1      0  71.2833         0
2         1       3    0  26.0      0      0   7.9250         2
3         1       1    0  35.0      1      0  53.1000         2
4         0       3    1  35.0      0      0   8.0500         2

## گام 2 - آموزش مدل با درخت تصمیم 

In [28]:
data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    1  22.0      1      0   7.2500         2
1         1       1    0  38.0      1      0  71.2833         0
2         1       3    0  26.0      0      0   7.9250         2
3         1       1    0  35.0      1      0  53.1000         2
4         0       3    1  35.0      0      0   8.0500         2

ستون‌های ۱ تا آخر: بردار ویژگی ما هستند

ستون صفرم: بردار هدف ما هست

In [29]:
X = data.iloc[:, 1:].values # Convert to numpy with values method
y = data.iloc[:, 0].values

In [30]:
X.shape, y.shape

((891, 7), (891,))

In [31]:
type(X), type(y) # numpy array type

(numpy.ndarray, numpy.ndarray)

با استفاده از کتابخانه، داده‌های خود را به دو قسمت داده‌های آموزشی و داده‌های تست تقسیم بندی می‌کنیم

هشتاد درصد داده‌های: آموزشی
بیست درصد داده‌ها: تست

random_state

در اینجا برای شبیه‌سازی یکسان تصادفی است

In [32]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=1234)

In [33]:
X_train.shape, y_train.shape

((712, 7), (712,))

In [34]:
X_test.shape, y_test.shape

((179, 7), (179,))

ساخت یک نمونه از جنگل تصادفی با تعداد معین از درخت تصمیم

In [35]:
rfc = RandomForestClassifier(n_estimators=100)

آموزش مدل با استفاده از داد‌های آموزشی

In [36]:
rfc.fit(X_train, y_train)

RandomForestClassifier()

تست کردن مدل با داد‌های آزمایشی

In [37]:
y_pred_test = rfc.predict(X_test)

In [38]:
y_pred_train = rfc.predict(X_train)

## گام 3 - گزارش عملکرد و نتایج مدل

محاسبه انواع گوناگون معیار‌های ارزیابی در مدل سازی طبقه‌بندی

ارزیابی مدل در نمونه‌های آموزشی

In [39]:
print(classification_report(y_train, y_pred_train))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99       440
           1       0.98      0.98      0.98       272

    accuracy                           0.98       712
   macro avg       0.98      0.98      0.98       712
weighted avg       0.98      0.98      0.98       712



ارزیابی مدل با نمونه‌های تست

In [40]:
print(classification_report(y_test, y_pred_test))

              precision    recall  f1-score   support

           0       0.84      0.88      0.86       109
           1       0.80      0.74      0.77        70

    accuracy                           0.83       179
   macro avg       0.82      0.81      0.82       179
weighted avg       0.83      0.83      0.83       179



In [41]:
confusion_matrix(y_test, y_pred_test)

array([[96, 13],
       [18, 52]])